lang/ir/src/ir/item_mod.rs

```diff
@@ -0,0 +1,616 @@
+            <ir::ItemMod as TryFrom<syn::ItemMod>>::try_from(item_mod)
+                .map_err(|err| err.to_string()),
+            Err(expected_err.to_string()),
+        );
+    }
+
+    #[test]
+    fn overlapping_messages_fails() {
+        assert_fail(
+            syn::parse_quote! {
+                mod my_module {
+                    #[ink(storage)]
+                    pub struct MyStorage {}
+
+                    impl MyStorage {
+                        #[ink(constructor)]
+                        pub fn my_constructor() -> Self {}
+
+                        #[ink(message, selector = "0xDEADBEEF")]
+                        pub fn my_message_1(&self) {}
+                    }
+
+                    impl MyStorage {
+                        #[ink(message, selector = "0xDEADBEEF")]
+                        pub fn my_message_2(&self) {}
+                    }
+                }
+            },
+            "encountered ink! messages with overlapping selectors (= [DE, AD, BE, EF])\n\
+                hint: use #[ink(selector = \"0x...\")] on the callable or \
+                #[ink(namespace = \"...\")] on the implementation block to \
+                disambiguate overlapping selectors.",
+        );
+    }
+
+    #[test]
+    fn overlapping_constructors_fails() {
+        assert_fail(
+            syn::parse_quote! {
+                mod my_module {
+                    #[ink(storage)]
+                    pub struct MyStorage {}
+
+                    impl MyStorage {
+                        #[ink(constructor, selector = "0xDEADBEEF")]
+                        pub fn my_constructor_1() -> Self {}
+
+                        #[ink(message)]
+                        pub fn my_message_1(&self) {}
+                    }
+
+                    impl MyStorage {
+                        #[ink(constructor, selector = "0xDEADBEEF")]
+                        pub fn my_constructor_2() -> Self {}
+                    }
+                }
+            },
+            "encountered ink! constructors with overlapping selectors (= [DE, AD, BE, EF])\n\
+                hint: use #[ink(selector = \"0x...\")] on the callable or \
+                #[ink(namespace = \"...\")] on the implementation block to \
+                disambiguate overlapping selectors.",
+        );
+    }
+
+    #[test]
+    fn allow_overlap_between_messages_and_constructors() {
+        assert!(
+            <ir::ItemMod as TryFrom<syn::ItemMod>>::try_from(syn::parse_quote! {
+                mod my_module {
+                    #[ink(storage)]
+                    pub struct MyStorage {}
+
+                    impl MyStorage {
+                        #[ink(constructor, selector = "0xDEADBEEF")]
+                        pub fn my_constructor() -> Self {}
+
+                        #[ink(message, selector = "0xDEADBEEF")]
+                        pub fn my_message(&self) {}
+                    }
+                }
+            })
+            .is_ok()
+        )
+    }
```

Comment:
Would be good to have a test with overlapping selectors from different traits with the same method, not just using `selector`

Reply:
i agree

Reply:
created an action item in the above ToDo list